src/memray/_memray/inject.cpp

```diff
@@ -0,0 +1,245 @@
+Memray_PyErr_ToString()
+{
+    std::string ret;
+
+    if (!PyErr_Occurred()) {
+        return ret;
+    }
+
+    PyObject* type;
+    PyObject* val;
+    PyObject* tb;
+    PyErr_Fetch(&type, &val, &tb);
+    PyErr_NormalizeException(&type, &val, &tb);
+
+    PyObject* exc_repr = PyObject_Repr(val);
+    if (!exc_repr) {
+        PyErr_Clear();
+        ret = "unknown exception (`repr(exc)` failed)!";
+    } else {
+        PyObject* utf8 = PyUnicode_AsUTF8String(exc_repr);
+        if (!utf8) {
+            PyErr_Clear();
+            ret = "unknown exception (`repr(exc).encode('utf-8')` failed)!";
+        } else {
+            ret = PyBytes_AsString(utf8);
+        }
+        Py_XDECREF(utf8);
+    }
+    Py_XDECREF(exc_repr);
+
+    Py_XDECREF(type);
+    Py_XDECREF(val);
+    Py_XDECREF(tb);
+
+    return ret;
+}
+
+bool
+run_script_impl(const std::string& script, std::string* errmsg)
+{
+    int rc;
+
+    PyObject* builtins = nullptr;
+    PyObject* globals = nullptr;
+    PyObject* code = nullptr;
+    PyObject* mod = nullptr;
+    bool success = false;
+
+    builtins = PyImport_ImportModule("builtins");
+    if (!builtins) {
+        goto done;
+    }
+
+    globals = PyDict_New();
+    if (!globals) {
+        goto done;
+    }
+
+    // Needed on 3.7 to avoid ImportError('__import__ not found')
+    rc = PyDict_SetItemString(globals, "__builtins__", builtins);
+    if (0 != rc) {
+        goto done;
+    }
+
+    code = Py_CompileString(script.data(), "_memray_attach_hook.py", Py_file_input);
+    if (!code) {
+        goto done;
+    }
+
+    mod = PyEval_EvalCode(code, globals, globals);
+    if (!mod) {
+        goto done;
+    }
+
+    success = true;
+
+done:
+    Py_XDECREF(mod);
+    Py_XDECREF(code);
+    Py_XDECREF(globals);
+    Py_XDECREF(builtins);
+
+    *errmsg = Memray_PyErr_ToString();
+    return success;
+}
+
+bool
+run_script(const std::string& script, std::string* errmsg)
+{
+    if (!Py_IsInitialized()) {
+        *errmsg = "Python is not initialized";
+        return false;
+    }
+
+    PyGILState_STATE gstate;
+    gstate = PyGILState_Ensure();
+    bool ret = run_script_impl(script, errmsg);
+    PyGILState_Release(gstate);
+    return ret;
+}
+
+void
+run_client(int arg)
+{
+    int port = (intptr_t)arg;
+    int sock = connect_client(port);
+    if (sock == -1) {
+        std::cerr << "memray attach failed!" << std::endl;
+        return;
+    }
+
+    std::string script;
+    if (!recvall(sock, &script)) {
+        std::cerr << "memray attach socket read error!" << std::endl;
+        return;
+    }
+
+    std::string errmsg;
+    bool success = run_script(script.c_str(), &errmsg);
+
+    if (!success) {
+        sendall(sock, errmsg);
+    }
+    ::close(sock);
+}
+
+extern "C" void*
+thread_body(void* arg)
+{
+    int rc = pthread_detach(pthread_self());
+    if (0 != rc) {
+        std::cerr << "Failed to detach thread!" << std::endl;
+    }
+
+    int port = reinterpret_cast<intptr_t>(arg);
```

Comment:
Isn't this narrowing from `intptr_t` to `int` implementation defined?

Reply:
Yes. That would be avoided by using an `unsigned short` instead - I'll change that.